<a href="https://colab.research.google.com/github/YACUERVO/AnalisisDatosFlores/blob/main/analisisMuestraEspeciesFlores.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [3]:
import csv
import math

#Cargar datos

def cargar_datos(filepath):
  with open (filepath , 'r') as file:
    cvs_reader = csv.reader(file)
    dataset = [row for row in cvs_reader if row] # Asegura excluir filas vacias
  return dataset


In [4]:
# Funcion permite agrupar por especie y sus diferentes caracteristicas

def agrupar_por_especie(datos):
  grupos = {}
  for fila in datos:
    especie = fila[-1]
    caracteristicas = list(map(float, fila[:-1])) #Convierte a float todas menos la especie
    if especie not in grupos:
      grupos[especie] = { 'Longitud_sepalo': [],
                          'ancho_sepalo': [],
                          'longitud_petalo': [],
                          'ancho_petalo': []}

    grupos[especie]['Longitud_sepalo'].append(caracteristicas[0])
    grupos[especie]['ancho_sepalo'].append(caracteristicas[1])
    grupos[especie]['longitud_petalo'].append(caracteristicas[2])
    grupos[especie]['ancho_petalo'].append(caracteristicas[3])
  return grupos




In [5]:
class Flor:

  def __init__(self, longitud_sepalo, ancho_sepalo, longitud_petalo, ancho_petalo):
      #Inicializa la flor
      self.longitud_sepalo = longitud_sepalo
      self.ancho_sepalo = ancho_sepalo
      self.longitud_petalo = longitud_petalo
      self.ancho_petalo = ancho_petalo

  def promedio(self, datos):
    #Calcula el promedio de una lista de valores
    return sum(datos) / len(datos)

  def desviacion_estandar(self, datos):
    #Calcula la desviacion estandar de una lista de valores
    promedio = self.promedio(datos)
    variance = sum([(x - promedio) ** 2 for x in datos]) / len(datos)
    return math.sqrt(variance)

  def analisis(self):
    "Devuelve el analisis de la flor promedio y desviacion estandar para cada caracteristica"

    analisis = {
        'Longitud_sepalo': {
            'promedio': self.promedio(self.longitud_sepalo),
            'desviacion_estandar': self.desviacion_estandar(self.longitud_sepalo)
        },
        'ancho_sepalo': {
            'promedio': self.promedio(self.ancho_sepalo),
            'desviacion_estandar': self.desviacion_estandar(self.ancho_sepalo)
        },
        'longitud_petalo': {
            'promedio': self.promedio(self.longitud_petalo),
            'desviacion_estandar': self.desviacion_estandar(self.longitud_petalo)
        },
        'ancho_petalo': {
            'promedio': self.promedio(self.ancho_petalo),
            'desviacion_estandar': self.desviacion_estandar(self.ancho_petalo)
        }
    }
    return analisis













In [9]:
#Mostrar datos
datos = cargar_datos('iris.data.txt')
print(datos)

#Agrupar por especie
flores = agrupar_por_especie(datos)
Iris_Setosa = Flor (flores['Iris-setosa']['Longitud_sepalo'],
                    flores['Iris-setosa']['ancho_sepalo'],
                    flores['Iris-setosa']['longitud_petalo'],
                    flores['Iris-setosa']['ancho_petalo'])

Iris_versicolor = Flor (flores['Iris-versicolor']['Longitud_sepalo'],
                    flores['Iris-versicolor']['ancho_sepalo'],
                    flores['Iris-versicolor']['longitud_petalo'],
                    flores['Iris-versicolor']['ancho_petalo'])


print(Iris_Setosa.analisis())
print(Iris_versicolor.analisis())


[['5.1', '3.5', '1.4', '0.2', 'Iris-setosa'], ['4.9', '3.0', '1.4', '0.2', 'Iris-setosa'], ['4.7', '3.2', '1.3', '0.2', 'Iris-setosa'], ['4.6', '3.1', '1.5', '0.2', 'Iris-setosa'], ['5.0', '3.6', '1.4', '0.2', 'Iris-setosa'], ['5.4', '3.9', '1.7', '0.4', 'Iris-setosa'], ['4.6', '3.4', '1.4', '0.3', 'Iris-setosa'], ['5.0', '3.4', '1.5', '0.2', 'Iris-setosa'], ['4.4', '2.9', '1.4', '0.2', 'Iris-setosa'], ['4.9', '3.1', '1.5', '0.1', 'Iris-setosa'], ['5.4', '3.7', '1.5', '0.2', 'Iris-setosa'], ['4.8', '3.4', '1.6', '0.2', 'Iris-setosa'], ['4.8', '3.0', '1.4', '0.1', 'Iris-setosa'], ['4.3', '3.0', '1.1', '0.1', 'Iris-setosa'], ['5.8', '4.0', '1.2', '0.2', 'Iris-setosa'], ['5.7', '4.4', '1.5', '0.4', 'Iris-setosa'], ['5.4', '3.9', '1.3', '0.4', 'Iris-setosa'], ['5.1', '3.5', '1.4', '0.3', 'Iris-setosa'], ['5.7', '3.8', '1.7', '0.3', 'Iris-setosa'], ['5.1', '3.8', '1.5', '0.3', 'Iris-setosa'], ['5.4', '3.4', '1.7', '0.2', 'Iris-setosa'], ['5.1', '3.7', '1.5', '0.4', 'Iris-setosa'], ['4.6', '

# 🌸 Análisis Estadístico de Flores por Especie (Python)

Este proyecto realiza un análisis estadístico básico (promedio y desviación estándar) de características florales, agrupadas por especie, usando programación orientada a objetos en Python. Es ideal como ejercicio introductorio al manejo de datos, clases y funciones estadísticas.

---

## 📌 Características Analizadas

Cada flor se describe mediante las siguientes características:

- Longitud del sépalo
- Ancho del sépalo
- Longitud del pétalo
- Ancho del pétalo
- Especie (nombre)

---

## 🧠 Objetivo del Proyecto

1. Agrupar flores por especie.
2. Calcular:
   - Promedio de cada característica.
   - Desviación estándar de cada característica.
3. Presentar un análisis estructurado por especie.

---

## 🛠️ Estructura del Proyecto

- `Flor`: clase que calcula estadísticas para un conjunto de flores.
- `agrupar_por_especie()`: función que agrupa los datos por especie.
- `main.py` o archivo de prueba: contiene datos de ejemplo y ejecución del análisis.

---

## 🧪 Ejemplo de uso

```python
from flor import Flor
from agrupador import agrupar_por_especie

# Lista de flores: 4 características + especie (último elemento)
datos = [
    [5.1, 3.5, 1.4, 0.2, 'Iris-setosa'],
    [4.9, 3.0, 1.4, 0.2, 'Iris-setosa'],
    [7.0, 3.2, 4.7, 1.4, 'Iris-versicolor']
]

# Agrupar por especie
grupos = agrupar_por_especie(datos)

# Analizar cada grupo
for especie, caracteristicas in grupos.items():
    flor = Flor(
        caracteristicas['Longitud_sepalo'],
        caracteristicas['ancho_sepalo'],
        caracteristicas['longitud_petalo'],
        caracteristicas['ancho_petalo']
    )
    print(f"\n📊 Análisis para {especie}")
    print(flor.analisis())